### PR TITLE
HSEARCH-3534 HSEARCH-3535  Default value  and out of bound message for "minimumShouldMatch"

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-3534
https://hibernate.atlassian.net/browse/HSEARCH-3535

### CODE REVIEW

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchBooleanJunctionPredicateBuilder.java [hunk 1]
```diff
@@ -124,6 +129,10 @@ protected JsonObject doBuild(ElasticsearchSearchPredicateContext context,
 		return outerObject;
 	}
 
+	private boolean hasAtLeastOneMustPredicate() {
+		return mustClauseBuilders != null || mustNotClauseBuilders != null;
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchBooleanJunctionPredicateBuilder.java]: From the Elasticsearch 6.7 documentation:

> If the bool query is in a filter context or has neither must or filter then at least one of the should queries must match a document for it to match the bool query. This behavior may be explicitly controlled by setting the minimum_should_match parameter.

https://www.elastic.co/guide/en/elasticsearch/reference/6.7/query-dsl-bool-query.html

The documentation mentions `must` and `filter`, but your condition mentions `must` and `mustNot`. I wonder if it's accurate? Maybe you should add tests (in the same test method) with "one should, one filter" and "one should, one mustNot" in a filter context, just to be sure?

Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchBooleanJunctionPredicateBuilder.java]: Good idea. Let's add the test so that to verify whether  `|| mustNotClauseBuilders != null` is really need.

Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchBooleanJunctionPredicateBuilder.java]: You were right, according to the new test. A must-not is treated differently from a must.

Comment (yrodiere) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchBooleanJunctionPredicateBuilder.java]: And what about "filter"? Or maybe you're still working on it?

Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchBooleanJunctionPredicateBuilder.java]: Sorry, I forgot it.
I think I've addressed it now.

**Comment by yrodiere** [comment]
Merged, thanks
